### DISCUSSION TITLE
Testing Dispose Directly With Usage Of "using" IDisposable

### DISCUSSION THREAD

**Discussion by Frueber** [ktmitton/Mittons.Fixtures#31] [Ideas]
I'm wondering if for tests that are verifying `Dispose`, and do indeed implement `IDisposable`, if we'd want to avoid usage of the `using` syntax as the compiler generates a good ol' newable wrapped in a try-finally and issues the `Dispose` call as well at the end of the block.
It likely might not hurt anything, but there is potential side effects with the double call.

**Comment by ktmitton** [comment]
I thought of that, but ultimately I figured it was better to continue doing "best practices", instead of throwing in a mix of when to do it "right" and when to do it "wrong".

**Reply by Frueber** [reply]
Interesting in the very direct testing context. The compiler does know of the double Dispose call as well and alerts to it. The using syntax is a "working practice" for most cases to simplify, but isn't always able to meet the correct scope or purpose depending on implementation. I do prefer it in pretty all cases though ha.
Here it ideally wouldn't cause unnecessary/unforeseen side effects if Dispose isn't doing something silly.

**Reply by ktmitton** [reply]
I was thinking about it more yesterday, and I think where I got to is in _this_ project, I always want to wrap our objects in usings, so any time somebody is looking at examples, they _know_ it's disposable, and they don't accidentally omit it in their code. Like, if we were building tests in a separate unrelated project that was using our objects like we were when you called it out, that's cool to omit using if they are testing dispose directly, because it's not their class, so people shouldn't be looking to them for guidance on how to use it, but since we define the class, we should always use it "properly".

**Reply by Frueber** [reply]
Would we consider closing its scope and confirming any side effects / outcomes in the end of the test? (Never explicitly calling Dispose, because that's not proper usage and they would want to copy that either?)

**Reply by ktmitton** [reply]
I think in this case, the point of the test is to test the function, so I want the test to be explicitly be doing that. Otherwise we're adding a dependency to the test of assuming the language will do what we expect. While in this case its probably a fair assumption, I think this carries over into other similar type test with frameworks that aren't baked into the language. I want to test my function so I know any framework using it behaves properly, I don't want to test my function _and_ the framework.

**Reply by ktmitton** [reply]
I'll be honest, this seems like a topic not great to talk about via text, we might be confusing each other. Let me know if you just want to talk about this in person sometime on a call to make sure we're both on the same page.

**Reply by Frueber** [reply]
Definitely up for a chat on it ☺️. 
It's also nothing we're needing any change to as of now so no hurry.

Your comment on the framework is partly why I'd think we'd want to be even more explicit with what we'd be testing. We "new up" the object and call the `Dispose` method. If there's extra "framework" in play we may convolute things, so reducing that might help express the intent. Always using `using` for an implementation of `IDisposable` isn't what the test is about.
It's just like any other implementation of an interface test.
The using could either be relied on as needing to work, or avoided in use during the explicit test of the `Dispose` method. Or, we do both as we were so that we have a secret intent of showing how using is a feature for implementations of `IDisposable`.

Just an interesting topic, nothing I'm hard-fast on.

**Comment by ktmitton** [comment]
@Frueber thought we'd move this over here so it can be talked about, and then we won't create an issue until we decide what action we want to take.

**Reply by Frueber** [reply]
Thanks! Wasn't sure where to chat about that.

**Reply by ktmitton** [reply]
I didn't originally have discussions enabled, it's a feature that was off by default...

I'm used to ADO, not github, so if there's something you see missing, odds are it's not by design, let me know and I can look into what needs to be added!

**Comment by ktmitton** [comment]
@Frueber so turns out this question goes away when everything gets converted to the IAsyncLifetime. Nothing is disposable now...

Are you good to close this?

**Reply by Frueber** [reply]
Oh very cool. The IAsyncLifetime seems to be a great fit.

**Reply by ktmitton** [reply]
I'm thinking I might change it a little, I really don't like xunit being a dependency, because up until now the fixture was generic for _any_ testing framework. I don't like that if you wanted to use another testing framework, you still need to take on the xunit dependency.

I'm thinking maybe we should create our own IAsyncLifetime interface, use that in project, and then create a separate package like Mittons.Fixtures.Xunit which adds something like XunitDockerEnvironmentFixture, which just wraps the normal fixture. Then you only have to pull in the xunit dependency if you want to.

**Reply by Frueber** [reply]
That's a good idea 💡. Will hopefully make it easier to guide those async constructions and disposes for non-xunit as well.

**Reply by ktmitton** [reply]
I created #57 to implement this. I'm probably not going to get to it right away, I want to focus on the health stuff, so feel free to pick it up if you want!

**Reply by ktmitton** [reply]
Once that's completed, I think then we'll be good to close this discussion
